src/javascript/waxeye.ts

```diff
@@ -354,58 +309,65 @@ interface ActionEval {
   type: ActionType.EVAL;
   exp: Expr;
   pos: number;
-  asts: Array<AST|string>;
+  asts: ASTList;
   err: RawError;
-  continuations: Continuation[];
+  continuations: ConsList<Continuation>;
 }
 
 function evalNext(
-    exp: Expr, pos: number, asts: Array<AST|string>, err: RawError,
-    continuations: Continuation[]): ActionEval {
+    exp: Expr, pos: number, asts: ASTList, err: RawError,
+    continuations: ConsList<Continuation>): ActionEval {
   return {type: ActionType.EVAL, asts, continuations, err, exp, pos};
 }
 
 interface ActionApply {
   type: ActionType.APPLY;
-  continuations: Continuation[];
+  continuations: ConsList<Continuation>;
   value: MatchResult;
 }
 
 function applyNext(
-    continuations: Continuation[], value: MatchResult): ActionApply {
+    continuations: ConsList<Continuation>, value: MatchResult): ActionApply {
   return {type: ActionType.APPLY, continuations, value};
 }
 
-function match(env: ParserConfig, start: keyof ParserConfig, input: string):
-    AST|ParseError {
+function match(
+    env: RuntimeParserConfig, start: keyof ParserConfig, input: string): AST|
+    ParseError {
   // move from initial state to halting state
   let action = moveEval(
       env, input,
-      evalNext(env[start].exp, 0, [], new RawError(0, [start], [], start), []));
+      evalNext(
+          env[start].exp, /*pos=*/0, /*asts=*/empty(),
+          new RawError(
+              /*pos=*/0, /*nonterminals=*/cons(start, empty()),
+              /*failedChars=*/empty(),
+              /*currentNT=*/start),
+          /*continuations=*/empty()));
   while (true) {
     switch (action.type) {
       case ActionType.EVAL:
         action = moveEval(env, input, action);
         break;
       case ActionType.APPLY:
         const {continuations, value} = action;
-        if (continuations.length === 0) {
+        if (continuations.isEmpty()) {
```

Comment:
Not a fan of using .isEmpty() and then .head and .tail. Why not use a switch like inhttps://github.com/basarat/typescript-book/blob/master/docs/types/discriminated-unions.md? I'm not sure if typescript gives exhaustiveness checks in this case though.

Ditto for similar occurrences later in the code.

Reply:
> Not a fan of using .isEmpty() and then .head and .tail.

This is type-safe, because `isEmpty()` is a [type guard](https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards). There is no field to discriminate for a `switch`.

Reply:
Oh I see, still looks a bit weird though. But that's probably a nit, otherwise code looks good to me.